### ISSUE TITLE
Named variaton for a model results in undefined method `new' for nil:NilClass

### ISSUE THREAD

**Issue by nrocco** [ytti/oxidized#1901]
As per the documentation here (https://github.com/ytti/oxidized/blob/master/docs/Creating-Models.md) it is mentioned that it is possible to create named variations for existing models. This is a great feature that I would like to use.

When following the documentation:

> Create a named variation of an existing model, by creating a file with a new name (such as ~/.config/oxidized/model/junos-extra.rb), Then require the original model and extend its base class as in the above example. The named variation can then be specified as an OS type for specific devices that can benefit from the extra functionality. This allows for preservation of the base functionality for the default model types.

I am getting this exception after running `oxidized`:

    /var/lib/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/manager.rb:17:in `load': undefined method `new' for nil:NilClass (NoMethodError)
        from /var/lib/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/manager.rb:59:in `loader'
        from /var/lib/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/manager.rb:47:in `add_model'
        from /var/lib/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/node.rb:170:in `resolve_model'
        from /var/lib/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/node.rb:22:in `initialize'
        from /var/lib/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/nodes.rb:21:in `new'
        from /var/lib/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/nodes.rb:21:in `block (2 levels) in load'
        from /var/lib/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/nodes.rb:16:in `each'
        from /var/lib/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/nodes.rb:16:in `block in load'
        from /var/lib/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/nodes.rb:130:in `synchronize'
        from /var/lib/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/nodes.rb:130:in `with_lock'
        from /var/lib/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/nodes.rb:10:in `load'
        from /var/lib/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/nodes.rb:125:in `initialize'
        from /var/lib/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/core.rb:14:in `new'
        from /var/lib/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/core.rb:14:in `initialize'
        from /var/lib/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/core.rb:4:in `new'
        from /var/lib/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/core.rb:4:in `new'
        from /var/lib/gems/2.3.0/gems/oxidized-0.26.3/lib/oxidized/cli.rb:13:in `run'
        from /var/lib/gems/2.3.0/gems/oxidized-0.26.3/bin/oxidized:8:in `<top (required)>'
        from /usr/local/bin/oxidized:23:in `load'
        from /usr/local/bin/oxidized:23:in `<main>'

Our setup includes:

- Ubuntu 16.04.6 LTS
- Ruby 2.3.1p112 (2016-04-26) [x86_64-linux-gnu]
- oxidized 0.26.3

This is the `router.db` file I use:

    $ cat ~/.config/oxidized/router.db
    router1:10.0.0.1:arista:oxidized:somepassword
    router2:10.0.0.2:arista-extra:oxidized:anotherpassword

This is the `config` file I use:

    $ cat ~/.config/oxidized/config
    ---
    interval: 3600
    log: /home/oxidized/.config/oxidized/logs2
    debug: true
    threads: 30
    timeout: 10
    retries: 0
    resolve_dns: false
    rest: 127.0.0.1:28888
    prompt: !ruby/regexp /^([\w.@-]+[#>]\s?)$/
    vars:
      auth_methods: [ "none", "publickey", "password", "keyboard-interactive" ]
      remove_secret: true
    groups: {}
    crash:
      directory: /home/oxidized/.config/oxidized/crashes
      hostnames: false
    input:
      default: ssh
      debug: true
      ssh:
        secure: false
    output:
      default: file
      file:
        directory: /home/oxidized/.config/oxidized/nodes
    source:
      default: csv
      csv:
        file: /home/oxidized/.config/oxidized/router.db
        delimiter: !ruby/regexp /:/
        map:
          name: 0
          ip: 1
          model: 2
          username: 3
          password: 4
    model_map:
      arista: eos
      arista-extra: eos-extra

The named variaton that defines `eos-extra`:

    $ cat ~/.config/oxidized/model/eos-extra.rb
    require 'oxidized/model/eos.rb'

    class EOS
      cmd 'show ip route | no-more' do |cfg|
        comment cfg
      end
    end

I do see the following debug messages in the oxidized logs just before the crash:

    D, [2019-09-14T09:52:06.266090 #18752] DEBUG -- : node.rb: resolving node key 'model', with passed global value of '' and node value 'eos-extra'
    D, [2019-09-14T09:52:06.266119 #18752] DEBUG -- : node.rb: setting node key 'model' to value 'junos' from global
    D, [2019-09-14T09:52:06.266145 #18752] DEBUG -- : node.rb: returning node key 'model' with value 'eos-extra'
    D, [2019-09-14T09:52:06.266167 #18752] DEBUG -- : lib/oxidized/node.rb: Loading model "eos-extra"
    D, [2019-09-14T09:52:06.266347 #18752] DEBUG -- : lib/oxidized/model/model.rb Added show ip route | no-more to the commands list
    F, [2019-09-14T09:52:06.266701 #18752] FATAL -- : Oxidized crashed, crashfile written in /home/oxidized/.config/oxidized/crash

The new command gets added to the commands list so the named variation file is found and read.

Are you able to help me point out what I am doing wrong here? I am more then willing to contribute to the documentation in this repository once I have a working setup.

Again, thank you so much for your efforts in creating and maintaining this awesome library!

**Comment by davama** [comment]
One thing i quickly see is rename eos-extra to eos.
You are not creating a new model, but extending the existing one. 
And on your source make sure you say eos

**Comment by nrocco** [comment]
I will try that.

But I expect that `router1` in my router.db gets backed up by my eos ruby file and that `router2` in my router.db gets backed up by the eos ruby file that ships with oxidized. 

I want to extend the build in eos model, but then *not* have that applied to all my devices running eos.

**Comment by ytti** [comment]
ACK we assume lot of stuff (convention over code principle) we assume class name based on file name and we try to instantiate class EOSExtra or something (unsure how we deal with dash as that is not valid class name), which doesn't exist.

Either make the file name eos or file name to eosextra and class to ruby```class EOSExtra < EOS```to inherit the EOS.

**Comment by davama** [comment]
> I want to extend the build in eos model, but then not have that applied to all my devices running eos.

You might have to create a custom eos model for this. Simply copy the contents of eos to your eosnew and update your source .

**Comment by ytti** [comment]
What OP want is perfectly possible. Just create new file, use same name class as the file, but require the original file and inherit its class.

file: pooper.rb
```ruby
require 'somemodel'
class Pooper < SomeModel
  ..
end
```

**Comment by davama** [comment]
> file: pooper.rb

👍 😆

**Comment by nrocco** [comment]
@davama @ytti thanks for your suggestions, however I still am not succeeding:

My router.db:

    $ cat ~/.config/oxidized/router.db
    router1:10.0.0.1:arista:oxidized:somepassword
    router2:10.0.0.2:aristaextra:oxidized:anotherpassword

My adapted part of the config:

    model_map:
        aristaextra: eosextra

And my model file:

    $ cat ~/.config/oxidized/model/eosextra.rb
    require 'oxidized/model/eos.rb'
    
    class EOSExtra < EOS
      cmd 'show ip route | no-more' do |cfg|
        comment cfg
      end
    end

Oxidized starts fine, my `router1` gets successfully backed up (using the build in eos.rb model) and my `router2` fails to back up:

```
D, [2019-09-15T10:33:31.607604 #7715] DEBUG -- : lib/oxidized/worker.rb: Added /10.0.0.2 to the job queue
D, [2019-09-15T10:33:31.607693 #7715] DEBUG -- : lib/oxidized/job.rb: Starting fetching process for 173.208.125.253 at 2019-09-15 10:33:31 UTC
D, [2019-09-15T10:33:31.607734 #7715] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2019-09-15T10:33:31.607864 #7715] DEBUG -- : lib/oxidized/job.rb: Config fetched for 173.208.125.253 at 2019-09-15 10:33:31 UTC
W, [2019-09-15T10:33:32.608226 #7715]  WARN -- : /10.0.0.2 status fail, retries exhausted, giving up
D, [2019-09-15T10:33:32.608470 #7715] DEBUG -- : lib/oxidized/worker.rb: Jobs running: 0 of 1 - ended: 2 of 2
D, [2019-09-15T10:33:32.608606 #7715] DEBUG -- : lib/oxidized/worker.rb: Running :nodes_done hook
```

I ran a `tcpdump` and couldn't seen any tcp traffic on port 22 towards `10.0.0.2`.

To rule out that host `10.0.0.2` I switched the value for the model field in my `router.db` and after running oxidized i am getting the same behaviour. `10.0.0.2` succeeds, `10.0.0.1` fails.

So the issue must be somewhere in this model extension logic.

The logs regaring loading of models are:

```
D, [2019-09-15T10:42:55.355408 #8024] DEBUG -- : lib/oxidized/node.rb: Loading model "eos"
D, [2019-09-15T10:42:55.355803 #8024] DEBUG -- : lib/oxidized/model/model.rb Added all to the commands list
D, [2019-09-15T10:42:55.355850 #8024] DEBUG -- : lib/oxidized/model/model.rb Added secret to the commands list
D, [2019-09-15T10:42:55.355879 #8024] DEBUG -- : lib/oxidized/model/model.rb Added show inventory | no-more to the commands list
D, [2019-09-15T10:42:55.355904 #8024] DEBUG -- : lib/oxidized/model/model.rb Added show running-config | no-more | exclude ! Time: to the commands list
```

and

```
D, [2019-09-15T10:42:55.415363 #8024] DEBUG -- : lib/oxidized/node.rb: Loading model "eosextra"
D, [2019-09-15T10:42:55.415557 #8024] DEBUG -- : lib/oxidized/model/model.rb Added show ip route | no-more to the commands list
```

Any other suggestions I can try to zoom in on the issue?

**Comment by ytti** [comment]
Hmm I'm not sure why that didn't work.

What if you do
```
EOSExtra = EOS
class EOSExtra
  ..
end
````

**Comment by nrocco** [comment]
Still no luck. With the following model file:

```
$ cat .config/oxidized/model/eosextra.rb
require 'oxidized/model/eos.rb'

EOSExtra = EOS

class EOSExtra
  cmd 'show ip route | no-more' do |cfg|
    comment cfg
  end
end
```

The end result is that both my `router1` and `router2` will have the `show ip route` command executed. But in my router.db file I only specifed `eosextra` for `router2`.

**Comment by ytti** [comment]
What about

EOSExtra = EOS.dup

**Comment by nrocco** [comment]
@ytti I tried your suggestion:

```
$ cat .config/oxidized/model/eosextra.rb
require 'oxidized/model/eos.rb'

EOSExtra = EOS.dup

class EOSExtra
  cmd 'show ip route | no-more' do |cfg|
    comment cfg
  end
end
```

But unfortunately I am getting the same behaviour as `EOSExtra = EOS`. Both `router1` and `router2` have the `show ip route` executed.

**Comment by nrocco** [comment]
> You might have to create a custom eos model for this. Simply copy the contents of eos to your eosnew and update your source.

@davama suggestion does give the expected behaviour.

**Comment by ytti** [comment]
I'd need to think about this more, I don't understand why subclassing and dup do not work. You shouldn't be modifying the existing EOS class at all. I don't currently understand the the outcome.

**Comment by nrocco** [comment]
@ytti Thanks for giving this some more time. Is there anything I can do to help you? Maybe zip and send you the contents of my `~/.config/oxidized` directory?

**Comment by davama** [comment]
> You might have to create a custom eos model for this. Simply copy the contents of eos to your eosnew and update your source.
@davama suggestion does give the expected behaviour.

Im curious about this inheriting thing. Going to give it a try myself.

**Comment by ytti** [comment]
Sorry @nrocco I've dropped the ball on this one.

Is this production only, or can you repro this in lab? If so, could we share like tmate or webex session where I look into this?

**Comment by nrocco** [comment]
@ytti No worries. I understand :)

We only reproduced this in our production environment (we do not have a lab or virtual EOS devices). I will dicuss with my team if it is possible to setup such a test lab quickly.

**Comment by nrocco** [comment]
@ytti We are currently in the process of creating a lab and reproducing the issue. Once we have reproduced this we can give you access to our lab so you can look around.

We will keep you posted.

**Comment by nrocco** [comment]
@ytti we have prepared a lab environment and we can give you ssh access to it. We reproduced the issue there.

Shall we get in contact via email? You can direct message me on twitter at `@nrocco`.

**Comment by ytti** [comment]
Email works, saku@ytti.fi

**Comment by davama** [comment]
@nrocco @ytti 
Does this mean https://github.com/ytti/oxidized/issues/1901#issuecomment-531477688 is now possible?
Thanks!

**Comment by ytti** [comment]
Yes. It was possible all the time, but required also 'cmd :ssh' to be defined on the child, which was a bug. So the model worked, but nothing happened, because no protocol was defined.

**Comment by nrocco** [comment]
Nice! Good to hear you made it work. 
We will wait for the next oxidized release so we can roll it out and get rid of some work arounds 👍

**Comment by ytti** [comment]
I can roll gem next week. You could of course just clone the repo and do 'rake build' to build the gem or use docker. The releases are in no way special from my side, they receive no testing and are arbitrarily released when someone wants new version to be gem installable, like now, which is fine, because it's been a while since that was done.

**Comment by ytti** [comment]
I just released 0.27.0.
